slurm_ops_manager/nvidia.py

```diff
@@ -0,0 +1,167 @@
+#!/usr/bin/env python3
+"""Nvidia Class."""
+
+import logging
+import os
+import shlex
+import subprocess
+from pathlib import Path
+
+from ops.framework import Object, StoredState
+from slurm_ops_manager.utils import operating_system
+
+
+logger = logging.getLogger()
+
+BASE_DIR = os.path.dirname(os.path.abspath(__file__))
+
+
+class NvidiaGPU(Object):
+    """Responsible for Nvdia GPU operations."""
+
+    _stored = StoredState()
+
+    def __init__(self, parent, key):
+        """Initialize class."""
+        super().__init__(parent, key)
+
+        self._stored.set_default(gpu_installed=False,
+                                 gpu_repo_configured=False,
+                                 gpu_repo_path="",
+                                 gpu_package_name="")
+
+        self._operating_system = operating_system()
+
+        # setup repositories based on official instructions
+        if self._operating_system == "ubuntu":
+            self._stored.gpu_repo_path = "/etc/apt/sources.list.d/nvidia_gpu.list"
+        elif self._operating_system == "centos":
+            self._stored.gpu_repo_path = "/etc/yum.repos.d/cuda-rhel7.repo"
+        else:
+            logger.error(f"#### Unsupported OS: {self._operating_system}")
+
+        # set a default package name and repository
+        if not self.package:
+            self.package = ""
+        if not self._stored.gpu_repo_configured:
+            self.repository = ""
+
+    @property
+    def installed(self) -> bool:
+        """Return wether nvidia driver is installed."""
+        return self._stored.gpu_installed
+
+    @property
+    def repository(self):
+        """Return the repository used for nvidia drivers."""
+        if self._stored.gpu_repo_configured:
+            return Path(self._stored.gpu_repo_path).read_text()
+        else:
+            return "Repository not set up"
```

Comment:
In general, I think it is best to indicate error conditions with exceptions. If you have a return type that indicates an error (empty string, None, False, etc), then you have to make sure that any code that is calling the function knows about the convention and handles it correctly. The error value convention might not be immediately obvious to another developer working on the code. This is especially true when the type of the error value is the same as the type that the function returns when it succeeds.

Here's a discussion about this: https://stackoverflow.com/questions/1152541/is-it-better-to-use-an-exception-or-a-return-code-in-python
Here's another discussion: https://softwareengineering.stackexchange.com/questions/405038/result-object-vs-throwing-exceptions

Reply:
> In general, I think it is best to indicate error conditions with exceptions.

I agree. But this is not the case here: we are not in an error state. There is no repository to return, so why raise an exception?

Another advantage is that is almost forces the user to check for errors, while when using exceptions this might be caught  somewhere in a generic way.